Engine/src/ECS/Systems/BaseSystem.bf

```diff
@@ -0,0 +1,318 @@
+using System.Collections;
+using SteelEngine.ECS.Components;
+using System;
+
+namespace SteelEngine.ECS.Systems
+{
+	/// <summary>
+	/// Defines the abstract classes for all <see cref="SteelEngine.ECS.Systems"/> classes.
+	/// </summary>
+	public abstract class BaseSystem
+	{
+		protected this(Application app)
+		{
+			App = app;
+			IsEnabled = true;
+			EntityToComponents = new Dictionary<EntityId, List<BaseComponent>>();
+			_uninitializedComponents = new Queue<BaseComponent>();
+			_entityRegistrationChecks = new List<EntityId>();
+
+			RegisterComponentTypes();
+		}
+
+		public Application App { get; protected set; }
+
+		/// <summary>
+		/// All tracked components.
+		/// </summary>
+		public Dictionary<EntityId, List<BaseComponent>> EntityToComponents ~ delete _;
+
+		/// <summary>
+		/// Whether or not the System has called <see cref="Initialize()"/>.
+		/// </summary>
+		public bool IsEnabled { get; set; }
+
+		/// <summary>
+		/// Whether or not the System has called <see cref="Initialize()"/>.
+		/// </summary>
+		public bool IsInitialized { get; protected set; }
+
+		/// <summary>
+		/// Entities that need registration checks.
+		/// </summary>
+		protected List<EntityId> _entityRegistrationChecks ~ delete _;
+
+		/// <summary>
+		/// Tracks all potentially uninitialized <see cref="SteelEngine.ECS.BaseComponent"/> objects.
+		/// All <see cref="SteelEngine.ECS.BaseComponent"/> objects will be initialized in the <see cref="Initialize"/>, <see cref="PreUpdate(GameTime)"/>, or <see cref="PostUpdate(GameTime)"/> methods.
+		/// </summary>
+		protected Queue<BaseComponent> _uninitializedComponents ~ delete _;
+
+		/// <summary>
+		/// Tracks all potentially uninitialized <see cref="SteelEngine.ECS.BaseComponent"/> objects.
+		/// All <see cref="SteelEngine.ECS.BaseComponent"/> objects will be initialized in the <see cref="Initialize"/>, <see cref="PreUpdate(GameTime)"/>, or <see cref="PostUpdate(GameTime)"/> methods.
+		/// </summary>
+		protected Type[] _registeredTypes ~ delete _;
+
+		/// <summary>
+		/// Adds a <see cref="SteelEngine.ECS.BaseComponent"/> to relevant <see cref="Components"/>. If the <see cref="SteelEngine.ECS.BaseComponent"/> is not initialized, it will be queued for initialization.
+		/// </summary>
+		/// <param name="componentToAdd"><see cref="SteelEngine.ECS.BaseComponent"/> to add.</param>
+		/// <returns>Whether or not the <see cref="SteelEngine.ECS.BaseComponent"/> was added to this BehaviorSystem's <see cref="Components"/>. Returns false if the <see cref="SteelEngine.ECS.BaseComponent"/> already existed.</returns>
+		protected virtual bool AddComponent(BaseComponent component)
+		{
+			let parent = component.Parent;
+			List<BaseComponent> entityComponents = ?;
+			if (parent == null)
+			{
+				return false;
+			}
+			bool isNewEntity = false;
+			if (!EntityToComponents.TryGetValue(parent.Id, out entityComponents))
+			{
+				isNewEntity = true;
+				entityComponents = new List<BaseComponent>();

```

Comment:
Debugger detects memory leak with trace from this line.

Reply:
The issue was happening when clearing empty entities. I was failing to delete the components list when removing the empty entity.